Source/Core/DolphinQt/Achievements/AchievementLeaderboardWidget.cpp

```diff
@@ -0,0 +1,118 @@
+#include <QGroupBox>
+#include <QLabel>
+#include <QLineEdit>
+#include <QPushButton>
+#include <QString>
+#include <QVBoxLayout>
+
+#include <fmt/format.h>
+
+#include <rcheevos/include/rc_api_runtime.h>
+#include <rcheevos/include/rc_api_user.h>
+#include <rcheevos/include/rc_runtime.h>
+
+#include "Common/CommonTypes.h"
+#include "Core/AchievementManager.h"
+#include "Core/Config/AchievementSettings.h"
+#include "Core/Config/MainSettings.h"
+#include "Core/Core.h"
+
+#include "DolphinQt/Config/ControllerInterface/ControllerInterfaceWindow.h"
+#include "DolphinQt/QtUtils/ModalMessageBox.h"
+#include "DolphinQt/QtUtils/NonDefaultQPushButton.h"
+#include "DolphinQt/QtUtils/SignalBlocking.h"
+#include "DolphinQt/Settings.h"
+
+AchievementLeaderboardWidget::AchievementLeaderboardWidget(QWidget* parent) : QWidget(parent)
+{
+  m_common_box = new QGroupBox();
+  m_common_layout = new QGridLayout();
+
+  {
+    std::lock_guard lg{*AchievementManager::GetInstance()->GetLock()};
+    UpdateData();
+  }
+
+  m_common_box->setLayout(m_common_layout);
+
+  auto* layout = new QVBoxLayout;
+  layout->setContentsMargins(0, 0, 0, 0);
+  layout->setAlignment(Qt::AlignTop);
+  layout->addWidget(m_common_box);
+  setLayout(layout);
+}
+
+void AchievementLeaderboardWidget::UpdateData()
+{
+  if (!AchievementManager::GetInstance()->IsGameLoaded())
+    return;
+  const auto& leaderboards = AchievementManager::GetInstance()->GetLeaderboardsInfo();
+  int row = 0;
+  for (const auto& board_row : leaderboards)
+  {
+    const AchievementManager::LeaderboardStatus& board = board_row.second;
+    QLabel* a_title = new QLabel(QString::fromStdString(board.name));
+    QLabel* a_description = new QLabel(QString::fromStdString(board.description));
+    QVBoxLayout* a_col_left = new QVBoxLayout();
+    a_col_left->addWidget(a_title);
+    a_col_left->addWidget(a_description);
+    m_common_layout->addLayout(a_col_left, row, 0);
+    // Each leaderboard entry is displayed with four values. These are *generally* intended to be,
+    // in order, the first place entry, the entry one above the player, the player's entry, and
+    // the entry one below the player.
+    // Edge cases:
+    // * If there are fewer than four entries in the leaderboard, all entries will be shown in
+    //   order and the remainder of the list will be padded with empty values.
+    // * If the player does not currently have a score in the leaderboard, or is in the top 3,
+    //   the four slots will be the top four players in order.
+    // * If the player is last place, the player will be in the fourth slot, and the second and
+    //   third slots will be the two players above them. The first slot will always be first place.
+    std::array<AchievementManager::LeaderboardRank, 4> to_display{};
+    for (size_t ix = 0; ix < to_display.size(); ix++)
+    {
+      to_display[ix] =
+          (board.entries.size() > ix) ? static_cast<AchievementManager::LeaderboardRank>(ix) : 0;
+    }
```

Comment:
This can just be replaced with `std::array<AchievementManager::LeaderboardRank, 4> to_display{0, 1, 2, 3};`... though I guess that's actually wrong and you want `{1, 2, 3, 4}`, no?

Reply:
(and in case you're wondering: No, the indices being possibly out of bounds do not matter, the `if (board.entries.count(rank) > 0)` check will return false for those anyway.)